src/Storages/AlterCommands.cpp

```diff
@@ -1465,14 +1465,6 @@ void AlterCommands::validate(const StoragePtr & table, ContextPtr context) const
                         ErrorCodes::BAD_ARGUMENTS,
                         "The change of data type {} of column {} to {} is not allowed. It has known bugs",
                         old_data_type->getName(), backQuote(column_name), command.data_type->getName());
-
-                bool has_object_type = isObject(command.data_type);
-                command.data_type->forEachChild([&](const IDataType & type){ has_object_type |= isObject(type); });
-                if (has_object_type)
-                    throw Exception(
-                        ErrorCodes::BAD_ARGUMENTS,
-                        "The change of data type {} of column {} to {} is not supported.",
```

Comment:
Should we check here that `old_data_type` is string and throw if it is not but `command.data_type` is object?

Reply:
No, because `String` is the only type that can be casted to `JSON` right now, so for other types the exception `TYPE_MISMATCH`